### DISCUSSION TITLE
How to include entire standard library in app

### DISCUSSION THREAD

**Discussion by jamesbcd** [pyinstaller/pyinstaller#6824] [Help] [answered]
My application allows for external python plugins. Though these plugins supply their own third-party dependencies, they assume that the entire python standard library is available.

Is there a simple way to ask pyinstaller to include the entire standard library when building my app, even if most of the library is not explicitly used/imported by my app?

**Comment by bwoodsend** [comment, marked as answer]
I really wouldn't recommend trying to make a PyInstaller application with plugins. It's one miserable developer environment for your plugin writers.

You can get a list of standard library modules either from [stdlibs](https://pypi.org/project/stdlibs/) or from [sys.stdlib_module_names](https://docs.python.org/3/library/sys.html#sys.stdlib_module_names) if you're happy to stick to Python >= 3.10. If you pass each of those through `collect_submodules()` then pass the output to `hiddenimports` in your spec file.
```python
from stdlibs import module_names
from PyInstaller.utils.hooks import collect_submodules

modules = []
for module in module_names:
    modules.extend(collect_submodules(module))

a = Analysis(
    ...,
    hiddenimports=modules,
    ...,
)

**Reply by jamesbcd** [reply]
Thanks, this works really well!

One extra question: when I used your technique above, `collect_submodules` didn't seem to always include the top-level module itself in the list it returned, and yet the top-level modules still seemed to end up in the resulting build just fine. The only thing I did differently to your suggestion was to attempt importing each module_name to see if it was actually importable. Would that be having an effect? Just trying to ensure I understand what's going on.
```python
from stdlibs import stdlib_module_names
from PyInstaller.utils.hooks import collect_submodules

exclusions = set(['antigravity','turtle','turtledemo','cmd','this','tkinter','idlelib','lib2to3',
                  'unittest','test','doctest','ensurepip',
                  'aifc','asynchat','asyncore','audioop','cgi','cgitb','chunk','crypt','imghdr','imp',
                  'mailcap','msilib','nis','nntplib','optparse','ossaudiodev','pipes','smtpd',
                  'sndhdr','spwd','sunau','telnetlib','uu','xdrlib'])
extra_module_names = []
for mod_name in sorted(stdlib_module_names()):
    try:
        if mod_name not in exclusions and mod_name[0] != "_":
            importlib.import_module(mod_name)
            # extra_module_names.append(mod_name) <-- I thought I would need this, but I didn't. Not sure why.
            submod_names = collect_submodules(mod_name)
            print("Including", mod_name, submod_names) # Example output: Including functools []
            extra_module_names.extend(submod_names)
    except (ModuleNotFoundError, ImportError) as e:
        print("Skipping", mod_name)

a = Analysis(
    ...,
    hiddenimports=extra_module_names,
    ...,
)
```

**Reply by bwoodsend** [reply]
Hmm, can you give me an example of a module where `module_name not in collect_submodules(module_name)`? That's not supposed to happen...

**Reply by rokm** [reply]
I think it happens when you pass it a module instead of a package; in that case, it returns an empty list. It might make sense to fix it so it returns the list with module name instead (and empty list if it is not importable, as we do now).

**Reply by jamesbcd** [reply]
Yes, when passed a module, `collect_submodules()` returns an empty list, but the module is still included in the build. So does `collect_submodules()` actually have the side-effect of modifying the build instructions, rather than just analysing packages?

**Reply by bwoodsend** [reply]
There's no magic like that involved. If they're being collected then they must be imported either by one of the other packages that `collect_submodules()` is picking up or by your own code (possibly indirectly by one of your dependencies).

**Reply by jamesbcd** [reply]
Okay, that makes more sense. So in my code above, it sounds like I should indeed uncomment this line:
```python
extra_module_names.append(mod_name)
```
That way the name of every module in the list will be passed to `hiddenimports` (in addition to the submodule names returned by `collect_submodules()`)?

**Reply by bwoodsend** [reply]
Until we fix that behaviour, yes, that would be sensible.
